shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -669,17 +687,33 @@ - (void)viewDidLoad {
   }
 
   if (@available(iOS 13.4, *)) {
-    _hoverGestureRecognizer.reset(
-        [[UIHoverGestureRecognizer alloc] initWithTarget:self action:@selector(hoverEvent:)]);
-    _hoverGestureRecognizer.get().delegate = self;
-    [_flutterView.get() addGestureRecognizer:_hoverGestureRecognizer.get()];
-
-    _panGestureRecognizer.reset(
-        [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(scrollEvent:)]);
-    _panGestureRecognizer.get().delegate = self;
-    _panGestureRecognizer.get().allowedScrollTypesMask = UIScrollTypeMaskAll;
-    _panGestureRecognizer.get().allowedTouchTypes = @[ @(UITouchTypeIndirectPointer) ];
-    [_flutterView.get() addGestureRecognizer:_panGestureRecognizer.get()];
+    _hoverGestureRecognizer =
+        [[UIHoverGestureRecognizer alloc] initWithTarget:self action:@selector(hoverEvent:)];
+    _hoverGestureRecognizer.delegate = self;
+    [_flutterView.get() addGestureRecognizer:_hoverGestureRecognizer];
+
+    _discretePanGestureRecognizer =
+        [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(scrollEvent:)];
+    _discretePanGestureRecognizer.allowedScrollTypesMask = UIScrollTypeMaskDiscrete;
+    _discretePanGestureRecognizer.allowedTouchTypes = @[];
+    _discretePanGestureRecognizer.delegate = self;
+    [_flutterView.get() addGestureRecognizer:_discretePanGestureRecognizer];
+    _continuousPanGestureRecognizer =
+        [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(panEvent:)];
+    _continuousPanGestureRecognizer.allowedScrollTypesMask = UIScrollTypeMaskContinuous;
+    _continuousPanGestureRecognizer.allowedTouchTypes = @[];
```

Comment:
why do we want to disallow all touch types? Does it mean that the callback is not called?

Reply:
When using trackpad scroll, it doesn't send touch events, instead it sends scroll events, so this doesn't stop the callback. If touch events are allowed, touching the screen will be consumed by the pan gesture recognizer instead of being passed through to flutter via `touchesBegan` etc.

Reply:
isn't trackpad scroll an "indirect" touch type?

https://developer.apple.com/documentation/uikit/uitouchtype/uitouchtypeindirect?language=objc

Reply:
That link is talking about the trackpad on Apple TV remote. If indirect touch events are allowed here, it means that clicking and dragging with ipad mouse or trackpad will be converted into scroll, which is not correct at all. This is tested, it only works as expected when all touch types are disallowed.

Reply:
strange. i think the tvOS is just an example and it is for iPadOS as well (as screenshot). Though I have not used this API yet. 

![Screen Shot 2022-06-06 at 2 54 38 PM](https://user-images.githubusercontent.com/41930132/172255711-44f55ce8-3888-47c6-99eb-323443ec71fb.png)

Reply:
Yes, the touches generated by tapping via iPad mouse and trackpad have `UITouchTypeIndirect`. But the scrolls are not touches at all, so this doesn't apply.